### PR TITLE
fix: client metrics name cache and unknown path handling (#3999)

### PR DESCRIPTION
**Pull request by tkrop**
### Description

This pull request is fixing two small bugs in the `apiUsageMonitoring` filter as well as improving the general test coverage of this filter:

1. It removes a small bug that prevented the caching of client metric names, which speeds up monitoring by 1.5us removing also 13 allocs per request as intended, and 
2. It fixes the originally intended discovery of application and API IDs in edge cases improving the api and application visible while making the code more transparent and readable.

### Related Issue(s)

fixes #3999.

### Testing

Run `make test`, since the code is highly unit tested.

### Checklist

- [x] I have reviewed and tested the changes thoroughly.
- [x] I have added appropriate documentation to the code for clarity.
- [x] I have communicated the consequences of the updates to customers.

Signed-off-by: Tronje Krop <tronje.krop@jactors.de>

### CODE REVIEW

**Comment by tkrop** [comment]
Not sure, why  `pr/tests` have been initially failing twice. For me it looks unrelated to the changes I have added. Locally it is timing out after 10 minutes in an unrelated package after having tested my package successfully.

File: filters/apiusagemonitoring/spec.go [hunk 1]
```diff
@@ -75,10 +75,12 @@ func NewApiUsageMonitoring(
...
 		log.Errorf(
-			"api-usage-monitoring-realmsTrackingPattern-tracking-pattern (global config) ignored: error compiling regular expression %q: %v",
+			"api-usage-monitoring-realmsTrackingPattern-tracking-pattern"+
+				" (global config) ignored: error compiling regular expression %q: %v",
 			realmsTrackingPattern, err)
 		realmsTrackingMatcher = regexp.MustCompile("services")
-		log.Warn("defaulting to 'services' as api-usage-monitoring-realmsTrackingPattern-tracking-pattern (global config)")
+		log.Warn("defaulting to 'services' as api-usage-monitoring-" +
+			"realmsTrackingPattern-tracking-pattern (global config)")
 	}
```
[the remark this answers is not in the document]
Comment (tkrop) [reply filters/apiusagemonitoring/spec.go:84]: While I agree that this is not a 100% clear behavior, it is still sound, since the cached `nil` is creating minimal impact while the missing error is just skipping repeated error messages. In a high performance system this suppression of repeating failure output may be even wanted.

@szuecs what is your take on this?

PS: I have a potential fix local, that caches result and error, but has the drawback of repeatedly logging the error - if not suppressed by the logging framework.

**Review by szuecs — APPROVED** [review]

**Comment by tkrop** [comment]
👍

**Review by MustafaSaber — APPROVED** [review]
